### PR TITLE
return default for failed api calls instead of crashing

### PR DESCRIPTION
**Pull request by skyqrose**
Deduplicates a lot of work writing and testing api calls.

Followup to https://github.com/mbta/skate/pull/415#discussion_r367615946

It's built on top of the branch for that PR, so until it merges, this PR will be a bit messy. The only new changes are in `api.ts` and `api.test.ts`.

### CODE REVIEW

**Comment by skyqrose** [comment]
Would calling it with named arguments in an object make it easier to understand the interface? like
```ts
export const fetchShapeForTrip = (tripId: TripId): Promise<Shape | null> =>
  apiCall({
    url: `/api/shapes/trip/${tripId}`,
    parser: (shape: Shape | null) => shape,
    defaultResult: null,
  })
```

**Comment by arkadyan** [comment]
That does feel like an improvement to me. Do you agree?

File: assets/src/api.ts [hunk 1]
```diff
@@ -45,6 +33,31 @@ const checkResponseStatus = (response: Response) => {
...
+    .then(parseJson)
+    .then(({ data: data }: { data: any }) => parser(data))
+    .catch(error => {
+      if (defaultResult === undefined) {
+        // tslint:disable-next-line: no-console
+        console.error(error)
+        throw error
+      } else {
+        // tslint:disable-next-line: no-console
+        console.warn(error)
```
Comment (arkadyan) [inline assets/src/api.ts]: We should listen to tslint and not call console here. Calling console is useless, and this is also doing what we want it to.

File: assets/tests/api.test.ts [hunk 2]
```diff
@@ -23,6 +24,80 @@ const mockFetch = (status: number, json: any): void => {
...
+
+    apiCall({
+      url: "/",
+      parser: () => null,
+    })
+      .then(() => {
+        done("fetchRoutes did not throw an error")
+      })
+      .catch(error => {
+        expect(error).not.toBeUndefined()
```
Comment (arkadyan) [inline assets/tests/api.test.ts]: There's also a [`.toBeDefined()`](https://jestjs.io/docs/en/expect#tobedefined)

**Pull request merged**